### ISSUE TITLE
unable to convert argument 3 from cdata<struct THCudaLongTensor*> to cdata<struct THCudaTensor*>

### ISSUE THREAD

**Issue by pby5** [torch/nn#1077]
my code is below:
require 'nn'; 
require 'cunn'
batch_size = 2
a = torch.FloatTensor({0.5, 0.5, 0.7, 0.3}):view(batch_size, -1):cuda()
b = torch.FloatTensor({1, 1}):cuda()
c = nn.ClassNLLCriterion():cuda()
d = nn.LogSoftMax():cuda()
e = nn.CrossEntropyCriterion():cuda()
c:forward(d:forward(a), b)
d:forward(a)  -- log softmax
e:forward(a, b)

and it exists an error:/home/gavinpan/torch/install/bin/lua: unable to convert argument 3 from cdata<struct THCudaLongTensor*> to cdata<struct THCudaTensor*>
stack traceback:
	[C]: in function 'v'
	/home/gavinpan/torch/install/share/lua/5.2/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
	...pan/torch/install/share/lua/5.2/nn/ClassNLLCriterion.lua:44: in function <...pan/torch/install/share/lua/5.2/nn/ClassNLLCriterion.lua:29>
	(...tail calls...)

why CudaLongTensor and I use CudaTensor?

**Comment by eriche2016** [comment]
maybe you need to update your torch, nn, cunn to the latest version.

**Comment by jonghyuk65** [comment]
I also have this problem... updating modules didn't help so I'm planning to reinstall everything anyway (not now lol)
In my case maxpooling and nll got error (maybe changed return type(cudatensor to cudalongtensor) of max function caused this)
I'm very new to torch so I was trying
https://github.com/soumith/cvpr2015/blob/master/Deep%20Learning%20with%20Torch.ipynb

In addition, how do I know version compatibility of torch modules? Always getting latest version is okay? There are so many modules and need to be synced (nn cunn cutorch ... )

**Comment by soumith** [comment]
@jonghyuk65 it is convenient to just go to cd ~/torch; ./update.sh; ./clean.sh; ./install.sh

**Comment by jonghyuk65** [comment]
@soumith Thanks! 
I tried this commands and it really seems to update my torch and installed packages.
But I still have the same problem... 
I tried to run train.lua in https://github.com/karpathy/char-rnn using gpu (no problem with cpu) and I got this messages.

/home/ubuntu/torch/install/bin/luajit: /home/ubuntu/.luarocks/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
        [C]: in function 'v'
        /home/ubuntu/.luarocks/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
        .../ubuntu/.luarocks/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'forward'
        train.lua:274: in function 'opfunc'
        /home/ubuntu/torch/install/share/lua/5.1/optim/rmsprop.lua:35: in function 'rmsprop'
        train.lua:314: in main chunk
        [C]: in function 'dofile'
        ...untu/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x004065d0

This char-rnn code seems to use nn for network and use optim to do rmsprop. 
I also think this is maybe just caused by version mismatch but I have no idea why updating modules doesn't work... I'm gonna try to find ways but I'm just leaving reply FYI.

**Comment by jonghyuk65** [comment]
For additional information
My installation status (result of luarocks list) of torch, cutorch, nn, cunn is
torch
   scm-1 (installed) - /home/ubuntu/.luarocks/lib/luarocks/rocks
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks
cutorch
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks
nn
   scm-1 (installed) - /home/ubuntu/.luarocks/lib/luarocks/rocks
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks
cunn
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks

and result of test.sh in torch folder also says that there are errors in GPU part.

...

  9/170 SoftShrink_backward ............................................. [PASS]
 10/170 ClassNLLCriterionSingleTarget ................................... [ERROR]
 11/170 SpatialFullConvolution_backward_batch ........................... [PASS]
 12/170 SpatialConvolutionLocal_forward_single .......................... [PASS]
 13/170 Sqrt_backward ................................................... [PASS]
 14/170 LogSigmoid_transposed ........................................... [PASS]
 15/170 ClassNLLCriterionMultipleTarget ................................. [ERROR]
 16/170 SpatialClassNLLCriterion ........................................ [ERROR]
 17/170 SoftMax_forward ................................................. [PASS]
 18/170 LogSoftMax_forward .............................................. [PASS]
 19/170 SpatialDilatedConvolution_forward_single ........................ [PASS]
 20/170 Threshold_backward .............................................. [PASS]
 21/170 distkldiv ....................................................... [PASS]
 22/170 SpatialAveragePooling_backward_batch ............................ [PASS]
 23/170 VolumetricConvolution_forward_single ............................ [ERROR]
 24/170 SoftPlus_backward ............................................... [PASS]
 25/170 TemporalConvolution_forward ..................................... [PASS]
 26/170 SpatialAveragePooling_includepad ................................ [PASS]
 27/170 Square_backward ................................................. [PASS]
 28/170 SpatialLogSoftMax_forward ....................................... [ERROR]
 29/170 BatchNormalization .............................................. [PASS]
 30/170 VolumetricFullConvolution_pair_test ............................. [PASS]
 31/170 LogSigmoid_backward ............................................. [PASS]
 32/170 VolumetricConvolution_backward_single ........................... [ERROR]
 33/170 Sqrt_transposed ................................................. [PASS]
 34/170 MarginCriterion_forward ......................................... [PASS]
 35/170 SparseLinear_backward ........................................... [PASS]
 36/170 TemporalConvolution_backward_batch .............................. [FAIL]

...

--------------------------------------------------------------------------------
ClassNLLCriterionSingleTarget
 Function call failed
/home/ubuntu/.luarocks/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
        [C]: in function 'v'
        /home/ubuntu/.luarocks/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
        .../ubuntu/.luarocks/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'forward'
        /home/ubuntu/torch/install/share/lua/5.1/cunn/test.lua:4029: in function 'v'
        /home/ubuntu/torch/install/share/lua/5.1/cunn/test.lua:5690: in function </home/ubuntu/torch/install/share/lua/5.1/cunn/test.lua:5688>
        [C]: in function 'xpcall'
        /home/ubuntu/.luarocks/share/lua/5.1/torch/Tester.lua:477: in function '_pcall'
        /home/ubuntu/.luarocks/share/lua/5.1/torch/Tester.lua:436: in function '_run'
        /home/ubuntu/.luarocks/share/lua/5.1/torch/Tester.lua:355: in function 'run'
        /home/ubuntu/torch/install/share/lua/5.1/cunn/test.lua:5711: in function 'testcuda'
        [string "nn.testcuda()"]:1: in main chunk
        [C]: in function 'pcall'
        ...untu/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
        [C]: at 0x004065d0

...

**Comment by soumith** [comment]
unfortunately, you have some local packages installed here: /home/ubuntu/.luarocks
I am not sure why you installed them there.
Do:

rm -rf /home/ubuntu/.luarocks

**Comment by jonghyuk65** [comment]
Some other guy installed torch before so I guess it's the reason... I
thought that it'd be okay because local package had same version. But
removing it solved my problem.

Thanks!

2016-12-27 0:23 GMT+09:00 Soumith Chintala <notifications@github.com>:
